File: Problems/leetcode/neetcode roadmap/arrays & hashing/encode_and_decode_strings.py
class Solution:

    def encode(self, strs: list[str]) -> str:
        s = ''
        for i in strs:
            s += f'{len(i)}#{i}'
        
        return s

    def decode(self, s: str) -> list[str]:
        strs = []

        i = 0
        while i < len(s):
            a = ''
            while s[i] != '#':
                a += s[i]
                i += 1

            i += 1
            strs.append(s[i:i+int(a)])
            i += int(a)
            
        
        return strs

File: Problems/leetcode/neetcode roadmap/arrays & hashing/encode_and_decode_strings.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:

    string encode(vector<string>& strs) {
        string ans = "";
        for(string s:strs)
        // int -> string = to_string()
        ans += to_string(s.size()) + "#" + s;
        return ans;
    }

    vector<string> decode(string s) {
        vector<string> ans;
        int i = 0;
        while(i != s.size()) {
            string num = "";
            while(s[i] != '#') {
                num += s[i];
                i++;     
            }
            
            // string -> int = stoi()
            int n = stoi(num);
            ans.push_back(s.substr(i+1, n));
            i+=n+1;
        }
        return ans;
    }
};
